src/state/sagas/workspace/SelectWorkspace/SelectWorkspace.js

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Cosmo Tech.
+// Licensed under the MIT license.
+
+import { call, fork, select, takeEvery, put } from 'redux-saga/effects';
+import { STATUSES } from '../../../commons/Constants';
+import { APPLICATION_ACTIONS_KEY } from '../../../commons/ApplicationConstants';
+import { SCENARIO_ACTIONS_KEY } from '../../../commons/ScenarioConstants';
+import { WORKSPACE_ACTIONS_KEY } from '../../../commons/WorkspaceConstants';
+import { SCENARIO_RUN_STATE } from '../../../../services/config/ApiConstants';
+import { getAllScenariosData } from '../../scenario/FindAllScenarios/FindAllScenariosData';
+import { fetchSolutionByIdData } from '../../solution/FindSolutionById/FindSolutionByIdData';
+import { getPowerBIEmbedInfoSaga } from '../../powerbi/GetPowerBIEmbedInfo/GetPowerBIEmbedInfoData';
+import { getFirstScenarioMaster } from '../../../../utils/SortScenarioListUtils';
+
+const getOrganizationId = (state) => state?.organization?.current?.data?.id;
+const selectSolutionIdFromCurrentWorkspace = (state) => state.workspace.current.data.solution.solutionId;
+const selectScenarioList = (state) => state.scenario.list.data;
+const getWorkspaces = (state) => state?.workspace?.list?.data;
+
+export function* selectWorkspace(selectedWorkspaceId) {
+  yield put({
+    type: APPLICATION_ACTIONS_KEY.SET_APPLICATION_STATUS,
+    status: STATUSES.LOADING,
+  });
+
+  const organizationId = yield select(getOrganizationId);
+  const workspaces = yield select(getWorkspaces);
+  const selectedWorkspace = workspaces && workspaces.find((workspace) => workspace.id === selectedWorkspaceId);
+
+  if (selectedWorkspace === undefined) {
+    yield put({
+      type: APPLICATION_ACTIONS_KEY.SET_APPLICATION_STATUS,
+      status: STATUSES.ERROR,
+      error: {
+        title: 'App initialization error',
+        status: null,
+        detail: `Could not find workspace with id ${selectedWorkspaceId} in workspaces list.`,
+      },
+    });
+    yield put({
+      type: WORKSPACE_ACTIONS_KEY.SET_CURRENT_WORKSPACE,
+      status: STATUSES.ERROR,
+      workspace: null,
+    });
+  }
+
+  yield put({
+    type: WORKSPACE_ACTIONS_KEY.SET_CURRENT_WORKSPACE,
+    status: STATUSES.SUCCESS,
+    workspace: selectedWorkspace,
+  });
+
+  yield call(getAllScenariosData, organizationId, selectedWorkspaceId);
+  const solutionId = yield select(selectSolutionIdFromCurrentWorkspace);
+  yield call(fetchSolutionByIdData, organizationId, selectedWorkspaceId, solutionId);
+  const scenarioList = yield select(selectScenarioList);
+  yield put({
+    type: SCENARIO_ACTIONS_KEY.SET_CURRENT_SCENARIO,
+    scenario: getFirstScenarioMaster(scenarioList), // Function returns null if list is empty
+    status: STATUSES.SUCCESS,
+  });
+  if (scenarioList.length !== 0) {
```

Comment:
can we use scenarioList.forEach here ?

Reply:
It looks like `yield` cannot be called inside a `forEach` block, but I found an equivalent syntax :slightly_smiling_face: